modules/snmp-dest/snmpdest.c

```diff
@@ -0,0 +1,899 @@
+  free(s->securityName);
+}
+
+static gboolean
+snmpdest_dd_session_init(SNMPDestDriver *);
+
+/* the max number of oid tokens */
+#define MAX_OIDS 128
+static gboolean
+snmpdest_worker_insert(SNMPDestDriver *self, LogMessage *msg, LogPathOptions *path_options)
+{
+  netsnmp_pdu *pdu;
+
+  msg_set_context(msg);
+  msg_set_context(NULL);
+
+  if (!self->session_initialized)
+    {
+      if (!snmpdest_dd_session_init(self))
+        {
+          msg_warning("SNMP: error in session init, message dropped",
+                      evt_tag_str("host", self->host),
+                      evt_tag_int("port", self->port),
+                      NULL);
+          log_msg_unref(msg);
+          return FALSE;
+        }
+    }
+
+  pdu = snmp_pdu_create(SNMP_MSG_TRAP2);
+  if (!pdu)
+    return FALSE;
+
+  gchar **oid_tokens;
+  oid parsed_oids[MAX_OIDS];
+  guint oid_cnt;
+  GList *snmp_obj = self->snmp_objs;
+  GList *snmp_code = self->snmp_codes;
+  GList *snmp_template = self->snmp_templates;
+  GString *fs = g_string_sized_new(128);
+  /* go through the snmp objects and add them to the message */
+  while (snmp_obj)
+    {
+      gchar *oid = snmp_obj->data;
+      if (oid[0] == '.')
+        {
+          /* skip the leading '.' in order not to parse an empty first token */
+          ++oid;
+        }
+
+      /* parse the oid string into oid values */
+      oid_tokens = g_strsplit(oid, ".", MAX_OIDS);
+      oid_cnt = 0;
+      while (oid_tokens[oid_cnt])
+        {
+          gint val;
+          sscanf(oid_tokens[oid_cnt], "%d", &val);
+          parsed_oids[oid_cnt] = val;
+          oid_cnt += 1;
+        }
+
+      g_strfreev(oid_tokens);
+
+      log_template_format(snmp_template->data, msg, &self->template_options, LTZ_LOCAL, 0, NULL, fs);
+
+      gint code = *((gint *)snmp_code->data);
+      /* check the number format and replace it with zero if necessary */
+      if (code == 0 || code == 1 || code == 3)
+        {
+          gboolean replace = FALSE;
+          gint i,
+               len = strlen(fs->str);
+          if (!len)
+            replace = TRUE;
+          else
+            for (i = 0; i < len && !replace; ++i)
+              if (fs->str[i] < '0' || fs->str[i] > '9')
+                replace = TRUE;
+          if (replace)
+            {
+              msg_warning("SNMP: invalid number replaced with '0'",
+                          evt_tag_str("value", fs->str),
+                          NULL);
+              g_string_assign(fs, "0");
+            }
+        }
+
+      gchar type_code = object_type_codes[code];
+      if (snmp_add_var(pdu, parsed_oids, oid_cnt, type_code, fs->str) != 0)
+        {
+          msg_warning("SNMP: error adding variable",
+                      evt_tag_str("value", fs->str),
+                      NULL);
+          log_msg_unref(msg);
+          return FALSE;
+        }
+
+      snmp_obj = snmp_obj->next->next->next; /* go to the next triplet and the corresponding items */
```

Comment:
Very optional: I would be tempted to turn this while loop into a macro (probably an anaphoric macro around for) :).

Reply:
I'll do this refactor alongside the threaded destination work